### DISCUSSION TITLE
Static Analysis Song

### DISCUSSION THREAD

**Discussion by 33577** [prosyslab-classroom/cs524-program-analysis#434] [Art competition]
## Song
정적 분석: 코드의 심연으로 
https://suno.com/song/551c5b4b-307a-486b-b4ef-01d4a5cbf43a

## Album cover
![image](https://github.com/prosyslab-classroom/cs524-program-analysis/assets/34697855/8e6af7c0-8597-42d0-9595-e22fd09b3b97)

## Lyrics
(Verse 1)
코드의 그림자, 버그가 숨어있네,
깊이 뛰어들어, 숨길 곳 없네.
Interval로 그리면, Sign으로 보이면,
미로 같은 코드 속, 멈추지 않으리.

(Chorus)
코드 심연으로 뛰어들어, 비밀을 찾아,
정적의 안개 속, 이유를 알아.
Interval이 묶고, Sign이 길을 비춰,
코드의 깊은 곳, 해답을 찾아.

(Verse 2)
Interval 범위, 무한대로 펼쳐지네,
가능성의 경계, 놓치지 않네.
Sign이 말해줘, 양수와 음수의 이야기,
이진 세계 속, 우리의 서사시.

(Bridge)
Congruence 속삭임, 주기의 비밀,
Octagon과 Polyhedra, 마음의 행렬.
모든 코드 줄, 탐험의 길,
끝없는 바다 속, 더 많은 걸 원해.

(Chorus)
코드 심연으로 뛰어들어, 비밀을 찾아,
정적의 안개 속, 이유를 알아.
Interval이 묶고, Sign이 길을 비춰,
코드의 깊은 곳, 해답을 찾아.

(Solo)

(Outro)
깊이를 두려워 마, 어둠을 두려워 마,
추상 도메인, 우리의 흔적.
논리의 미로 속, 모든 사슬을 끊어,
분석의 힘으로, 모든 걸 정복해.

(Chorus)
코드 심연으로 뛰어들어, 비밀을 찾아,
정적의 안개 속, 이유를 알아.
Interval이 묶고, Sign이 길을 비춰,
코드의 깊은 곳, 해답을 찾아.

(Outro)
코드 심연으로 뛰어들어, 절대 쓰러지지 않아,
우리의 분석으로, 모두 정복하리.

## Name
전종욱

## Tool You Used
ChatGPT-4o, Suno 

## Short Description
저한테, [정적 분석]은 means [요약 해석으로 모르던 코드의 심연을 정복하는 것]입니다.

## Long Description (up to 3 sentences)
어떤 코드에 버그가 있는지 알기란 어렵습니다. 하지만 정적 분석을 통해 이를 발견할 수 있습니다. 
이 노래는 Interval, Sign, 또는 다른 요약 도메인을 사용한 분석이 코드를 이해하고 정복하는 데 어떻게 도움이 되는지에 대한 내용입니다.

(English)

## Song
Static Analysis: Under the code
https://suno.com/song/551c5b4b-307a-486b-b4ef-01d4a5cbf43a

## Name
Jonguk Jeon

## Tool You Used
ChatGPT-4o, Suno 

## Short Description
For me, [Static Analysis] means [Conquering the unknown code through abstract interpretation]

## Long Description (up to 3 sentences)
It is hard to know whether some code has bugs or not. However, with static analysis, we can uncover them. This song is about how analysis using intervals, signs, or other abstract domains helps to understand and conquer the code.

**Comment by KihongHeo** [comment]
Oh. I think we need Produce 101.

**Comment by KihongHeo** [comment]
우리 추상이라는 혼란스러운 말 대신 요약이라고 쓰는게 어떨까요?
한국어에서 추상적이라고 하면 어떤 느낌이 드나요? (예: "종욱씨 말이 참 추상적이에요")
Djiskstra가 말씀하셨지요.
> “The purpose of abstraction is not to be vague, but to create a new semantic level in which one can be absolutely precise.” -

**Reply by 33577** [reply]
교수님 제안을 듣고 "Abstraction", "추상", "요약"의 사전적 정의를 찾아보았습니다. 생각해볼 기회를 주셔서 감사합니다. 
저는 추상이란 단어가 "Abstraction"을 잘 표현한다고 생각합니다. 추상의 사전적 정의는 "Abstraction"의 사전적 정의와 꽤 유사합니다. 

들어주신 예문은 부정적인 맥락에서 사용되었기에 부정적으로 들립니다. 
"OO씨는 추상적인 생각도 잘 표현하네요"라고 하면 긍정적인 뜻으로 들립니다. 

요약은 저에겐 좀 어색하게 느껴지는데, 사전적 정의에서 그렇듯 그 대상이 말이나 글로 한정되기 때문인 것 같습니다. 
그래서 "사용자의 주문을 추상화하면", "메시지 발행기를 추상화하면"과 같은 말은 자연스러우나 
"사용자의 주문을 요약하면", "메시지 발행기를 요약하면"은 어색하거나 원하는 의도처럼 들리진 않는 것 같습니다. 

과도하게 모호하거나 어려운 단어는 지양해야 한다고 생각합니다. 
그렇지만 의미를 정확하게 나타내는 것이 쉽게 말하는 것보다 더 중요하다고 생각합니다. 

그래도 일관된 용어를 사용하는 것 또한 중요하기에 이 수업에서 요약을 공식적으로 채택하기로 한다면, 그대로 수정하겠습니다 :) 


> Abstraction - oxford dictionary
a general idea not based on any particular real person, thing or situation

> Abstraction - wikipedia
Abstraction is a process wherein general [rules](https://en.wikipedia.org/wiki/Rule_of_inference) and [concepts](https://en.wikipedia.org/wiki/Concept) are derived from the usage and classification of specific examples, literal ([real](https://en.wikipedia.org/wiki/Reality) or [concrete](https://en.wikipedia.org/wiki/Abstract_and_concrete)) signifiers, [first principles](https://en.wikipedia.org/wiki/First_principle), or other methods.

> 추상 - 표준국어대사전
여러 가지 사물이나 개념에서 공통되는 특성이나 속성 따위를 추출하여 파악하는 작용.

> 요약 - 표준국어대사전
말이나 글의 요점을 잡아서 간추림.

**Reply by KihongHeo** [reply]
네. 사전적 정의에만 따라서 하는 것은 기계적인 번역이고 우리가 지양해야할 바 입니다. 해당 언어에서 쓰임새를 잘 고려할 필요가 있겠지요.

"추상적인것도 잘 표현하네요"에서 추상적은 여전히 부정적인 말 아닌가요? 추상적임에도 "잘" 한다는 거니까. "잘" 이 긍정일 뿐이지요.

추상이라는 말은 일반인이 부정적인 뉘앙스 이외엔 잘 쓰지 않는데요. 우리가 일상적으로 잘 쓰는 말로 번역하는 것이 중요합니다. 그래야 비전문가나 처음 배우는 사람들도 쉽게 접할 수 있거든요.

**Reply by 33577** [reply]
아 사전적 정의를 맹목적으로 따라하는 것을 지양해야 할 수 있군요. 
적절한 번역을 찾으려고 국립국어원에 한 문의를 기다리느라 다소 늦게 답신 드립니다. 기다림이 무색하게도 좋은 답변은 못 받았습니다. 

저는 "추상"이란 단어가 일반인에게도 꽤 친숙한 단어라고 생각합니다. 미술계에서도 비슷한 뜻으로 사용하고 있고, 일상적인 픽토그램 등에서도 사용됩니다. 

그래도 더 쉬운 수준으로 순화한 말이 필요하다고 한다면, "요약"을 포함하여 "일반화", "단순화", "간소화", "추리기", "간추리기", "핵심 추출하기" 등 다양한 단어를 떠올릴 수 있을 것 같습니다. 저는 이 중 "핵심 추출하기"나 "간추리기"가 좋은 것 같습니다.

- "일반화", "단순화", "간소화"
영문에서 "Abstraction"을 "Generalization"이나 "Simplification"과 구별해 사용하는 점에서 "일반화", "단순화", "간소화"는 적절하지 않은듯 합니다.
- "추리기"
"추리기"는 일상에서 자주 사용하는 단어가 아니라 제외했습니다. 
- "간추리기"
"간추리기"는 "요약"처럼 말이나 글에 대해 한정되는 듯한 느낌을 줍니다만 제게는 "요약"보다 자연스러운 것 같습니다. 
  GPT4o도 "간추리기"를 좀 더 추천합니다. 
```
// Chat-GPT
"간추리기"와 "요약하기"는 비슷하지만 미묘한 차이가 있습니다.
- 간추리기:
의미: 중요한 요소나 핵심 사항만을 추려내어 간결하게 만드는 것.
사용 맥락: 주로 복잡한 내용에서 중요하고 핵심적인 부분만 골라내는 경우에 사용.
예: 프로그래밍에서 복잡한 코드 구조를 간추려서 이해하기 쉽게 만드는 과정.

- 요약하기:
의미: 전체 내용을 간단하게 줄여서 핵심적인 내용만을 전달하는 것.
사용 맥락: 긴 문서나 이야기를 짧게 줄이는 경우에 사용.
예: 긴 보고서를 요약하여 핵심 내용을 간략하게 전달하는 과정.
이 두 단어는 비슷한 의미로 사용될 수 있지만, "간추리기"는 복잡한 구조에서 중요한 부분을 추려내는 느낌이 더 강하고, 
"요약하기"는 전체 내용을 간단하게 줄이는 느낌이 더 강합니다. 
컴퓨터 과학에서 "Abstraction"을 표현할 때는 "간추리기"가 더 적합할 수 있습니다.
```
- "핵심 추출하기"
의미를 잘 표현하는 것 같으나 다소 장황합니다.

**Reply by KihongHeo** [reply]
네. 좋습니다. 마구 시도해보세요. 사전, 국어원이나 GPT가 시키는대로 해야할 것이 아니라 우리가 만들어가고 그들이 따라오게 해야합니다. 왜냐면 이 분야는 우리가 전문가이니까요.

**Reply by 33577** [reply]
아 그렇군요 외부 자료는 참고만 하고 주체적으로 더 고민해보겠습니다. 생각하지 못했던 관점 알려주셔서 감사드립니다 :)
